### DISCUSSION TITLE
endpoint to file upload

### DISCUSSION THREAD

**Discussion by avelino** [prest/prest#553] [API Server]
**Is your feature request related to a problem? Please describe.**
APIs are generally not designed for data transfer, it is common to need to work with file uploads.

Today it is not possible to deal with file uploads - unless you have a blob field in the database (not a good practice).

**Describe the solution you'd like**
Endpoint that receives file (base64 and file) and saves to object store (_example_ `s3`): `/upload`

This endpoint is to receive the file and return the hosting url (public or private), it will not save to the database.
The job of saving to the database will be done through another request to the existing pRest endpoint.

**Workflow**

1. POST: request `image upload`
2. response `image upload`
3. POST: request `/{DATABASE}/{SCHEMA}/{TABLE}` - one fields contains url of the image that was uploaded
4. response `/{DATABASE}/{SCHEMA}/{TABLE}`

**Additional context**
s3 api support, but with adapter architecture, so we give room to develop new adapters

**Comment by fabriziomello** [comment]
Not sure if we'll get too much benefit implementing it because the steps 1 and 2 can be done easily by external object store CLI from each cloud vendor.

Another thing that came in my mind is that nowadays pREST requests are atomic (cause the backend is a RDBMS ofc). So if we add the notion of "file upload" as described in this issue we'll break it completely, and there are no easy way to guarantee it. Think about the steps 1 and 2 complete with success and step 3 fail... so the client application should deal with it.

@avelino is this a real feature to be incorporated to the core?? or maybe it should be material to a plugin??

**Comment by avelino** [comment]
@fabriziomello good point, 

Usually, software in production will not use the CLI from the cloud provider, but have to implement a new endpoint for image upload. I will go further, usually real APIs do not have a single endpoint for image upload. If you have an image field, you send a request to the endpoint (`/xyz`) containing the data + file (base64) and it uploads it to the object store and with the returned result saves the public (or private) url in the database table.

We can develop an external service (new software or use some existing software) to handle "just" image uploads.

The problem is that whoever is consuming the API will have no authentication contract (tokens) because it is "not possible" to share between provider and other service.

I see this as an optional feature that by default comes deactivated (endpoint does not exist) and when activated prest delivers this new endpoint

**Comment by fabriziomello** [comment]
> @fabriziomello good point,
> 
> Usually, software in production will not use the CLI from the cloud provider, but have to implement a new endpoint for image upload. 

Sure... what I really meant is use the related API (from cloud vendors) to interact with Object Store.

> I will go further, usually real APIs do not have a single endpoint for image upload. If you have an image field, you send a request to the endpoint (`/xyz`) containing the data + file (base64) and it uploads it to the object store and with the returned result saves the public (or private) url in the database table.

How we'll guarantee the request atomicity??? It's not simple ... What I already saw out there is implementations using the RDBMS as a staging storage and in background another service sending asynchronously to external Object Store and removing from RDBMS saving the corresponding URL

> We can develop an external service (new software or use some existing software) to handle "just" image uploads.
> 

Plugin?

> The problem is that whoever is consuming the API will have no authentication contract (tokens) because it is "not possible" to share between provider and other service.
> 

Not sure how to solve it...


> I see this as an optional feature that by default comes deactivated (endpoint does not exist) and when activated prest delivers this new endpoint

I'm in favor to threat it as a plugin and first solve the #466 .

**Reply by avelino** [reply]
I like the idea of plugins, I still have no idea of the work, but I will do a POC to understand the complexity (evolve this game #466)

I will start with option to create plugin to create custom endpoints

**Comment by DiegoSantosWS** [comment]
I like the idea of this implementation being in plugin format.

That way we would be able to make implementations with other coulds, such as DO and GCP Storage.

Using the idea of enabling the `endpoint` the user (client) can inform the cloud and the access key/secret at format `base64`.

**Comment by avelino** [comment]
> ⚠️  I have converted the issue into a discussion because we have nothing defined of how to do it, but some ideas, as soon as we reach a common point of how we will proceed I will open a new issue and mark this discussion.
